File: frontend/types/sync-run.ts
export type SyncRunStatus = "queued" | "running" | "completed" | "failed";
export type SyncRunMode = "quick" | "full";

export type SyncRunStage =
  | "queued"
  | "reading_sheet"
  | "saving_videos"
  | "analyzing_titles"
  | "processing_transcripts"
  | "finding_hook_patterns"
  | "syncing_comments"
  | "finalizing";

export interface SyncRunWarning {
  code: string;
  detail: string;
}

export interface LastSyncStatus {
  run_id: number;
  mode: SyncRunMode;
  finished_at: string;
  duration_seconds: number | null;
  /** Unique videos in DB (matches dashboard catalog). */
  catalog_video_count: number;
  /** Rows read from Google Sheets (may exceed catalog due to duplicates). */
  sheet_rows: number;
  warning_count: number;
}

export interface SyncRunResult {
  mode?: SyncRunMode;
  total_rows?: number;
  created?: number;
  updated?: number;
  titles_analyzed?: number;
  transcripts_processed?: number;
  transcript_text_from_sheet?: number;
  hook_patterns_found?: number;
  audience_discussions_added?: number;
  warning_count?: number;
  warnings?: SyncRunWarning[];
  duration_seconds?: number;
  embeddings_created?: number;
  transcripts_fetched?: number;
  comments_fetched?: number;
  hooks_indexed?: number;
  catalog_video_count?: number;
  sheet_rows?: number;
}

export interface SyncRun {
  id: number;
  mode: SyncRunMode;
  status: SyncRunStatus;
  stage: string;
  processed: number;
  total: number | null;
  message: string | null;
  current_entity_name: string | null;
  warning_count: number;
  warnings: SyncRunWarning[];
  result: SyncRunResult | null;
  error_message: string | null;
  started_at: string;
  finished_at: string | null;
  duration_seconds: number | null;
}

export interface SyncRunStartResponse {
  run_id: number;
  status: SyncRunStatus;
  mode: SyncRunMode;
}

export function isSyncRunActive(run: SyncRun | null): boolean {
  return run?.status === "queued" || run?.status === "running";
}

export function syncRunProgressPercent(run: SyncRun): number | null {
  if (run.total == null || run.total <= 0) return null;
  return Math.min(100, Math.round((run.processed / run.total) * 100));
}

File: backend/app/schemas/sync_run.py
"""Schemas for background Sheets sync runs."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

SyncRunStatus = Literal["queued", "running", "completed", "failed"]
SyncRunMode = Literal["quick", "full"]
SyncRunStage = Literal[
    "queued",
    "reading_sheet",
    "saving_videos",
    "analyzing_titles",
    "processing_transcripts",
    "finding_hook_patterns",
    "syncing_comments",
    "finalizing",
]


class SyncStartRequest(BaseModel):
    mode: SyncRunMode = "quick"


class SyncRunStartResponse(BaseModel):
    run_id: int
    status: SyncRunStatus = "queued"
    mode: SyncRunMode = "quick"


class SyncRunRead(BaseModel):
    id: int
    mode: SyncRunMode = "full"
    status: SyncRunStatus
    stage: str
    processed: int = 0
    total: int | None = None
    message: str | None = None
    current_entity_name: str | None = None
    warning_count: int = 0
    warnings: list[dict[str, Any]] = Field(default_factory=list)
    result: dict[str, Any] | None = None
    error_message: str | None = None
    started_at: datetime
    finished_at: datetime | None = None
    duration_seconds: int | None = None

    model_config = {"from_attributes": True}


class LastSyncStatus(BaseModel):
    """Latest completed sync for UI badges."""

    run_id: int
    mode: SyncRunMode
    finished_at: datetime
    duration_seconds: int | None = None
    catalog_video_count: int = 0
    sheet_rows: int = 0
    warning_count: int = 0
    relative_label: str | None = None


class SyncRunSummary(BaseModel):
    """Human-facing completion stats (also stored in result_json)."""

    total_rows: int = 0
    created: int = 0
    updated: int = 0
    titles_analyzed: int = 0
    transcripts_processed: int = 0
    transcript_text_from_sheet: int = 0
    hook_patterns_found: int = 0
    audience_discussions_added: int = 0
    warning_count: int = 0
    duration_seconds: int | None = None
